Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,835 @@
+        let newHeader = FrameHeader(length: bytes.readableBytes, type: originalHeader.type,
+                                    flags: originalHeader.flags | HTTP2Frame.FrameFlags.endHeaders.rawValue,
+                                    rawStreamID: originalHeader.rawStreamID)
+        return try self.readFrame(withHeader: newHeader, from: &bytes)      // pretend it's one big HEADERS or PUSH_PROMISE frame
+    }
+    
+    private mutating func readPotentialFrame(from bytes: inout ByteBuffer) throws -> HTTP2Frame? {
+        assert(self.isIdle)
+        guard let header = bytes.readFrameHeader() else {
+            return nil
+        }
+        
+        if bytes.readableBytes >= header.length {
+            // we can read the entire thing already
+            return try self.readFrame(withHeader: header, from: &bytes)
+        } else if header.type != 0 /* DATA frame */ {
+            // start accumulating bytes and update our state
+            var buffer = allocator.buffer(capacity: header.length)
+            buffer.write(buffer: &bytes)    // consumes the bytes from the input buffer
+            self.state = .accumulatingData(header, buffer)
+            return nil
+        } else {
+            if bytes.readableBytes == 0 {
+                // we only have the header of a DATA frame. Don't emit a frame, but keep the header around
+                // to synthesize more once we have the data.
+                if header.flags & 0x08 != 0 {
+                    // it's padded. We can't enter .simulatingDataFrames without knowing our pad length, so
+                    // work around this by un-reading the header bytes and returning nil.
+                    bytes.moveReaderIndex(to: bytes.readerIndex - 9)
+                } else {
+                    // no padding to read, so everything to follow is payload: just enter simulation state
+                    // with all bytes remaining.
+                    self.state = .simulatingDataFrames(originalHeader: header, padding: 0, remainingBytes: header.length)
+                }
+                
+                return nil
+            }
+            
+            // Received a partial DATA frame: return current bytes as a single frame, then simulate
+            // further frames as more of the payload arrives
+            let streamID = HTTP2StreamID(knownID: Int32(header.rawStreamID & ~0x8000_0000))
+            guard streamID != .rootStream else {
+                // DATA frames MUST be associated with a stream. If a DATA frame is received whose
+                // stream identifier field is 0x0, the recipient MUST respond with a connection error
+                // (Section 5.4.1) of type PROTOCOL_ERROR.
+                throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+            }
+            
+            var slice = bytes
+            let consumed = slice.readableBytes
+            bytes.clear()       // consumed all the bytes
```

Comment:
I think this operation is a fancy way to spell `var slice = bytes.readSlice(length: bytes.readableBytes)`

Reply:
In fact, it's worse than that as `clear` will definitely CoW in this mode.